### ISSUE TITLE
[RFC]: Force symbolic shapes in `torch_binding_meta.py`

### ISSUE THREAD

**Issue by yiz-liu** [vllm-project/vllm-ascend#11006]
### Motivation.

`csrc/torch_binding_meta.cpp` contains meta implementations for vLLM Ascend custom ops. These functions are used by PyTorch fake tensor mode, `torch.compile`, export, and graph capture. A meta kernel must describe output shape, dtype, and device without running the real NPU kernel.

Several meta kernels currently read tensor dimensions through concrete APIs:

```cpp
int64_t batch = input.size(0);
auto out = at::empty({batch, hidden}, input.options());
```

This drops PyTorch symbolic shape information. Under `torch.compile(dynamic=True)`, the first dimension may be a `SymInt` such as `s0`. If the meta kernel converts it to `int64_t`, the FX graph records the first observed batch size instead of `s0`.

The correct pattern is:

```cpp
auto batch = input.sym_size(0);
c10::SymDimVector shape = {batch, c10::SymInt(hidden)};
auto out = at::empty_symint(shape, input.options());
```

## Why This Matters in vLLM

vLLM compiles model graphs from dummy inputs before serving real requests. Those dummy inputs often use scheduler limits such as `max_num_batched_tokens`, while real serving batches vary by request mix, prefill length, decode length, and speculative decoding state.

If a custom op meta kernel turns symbolic dimensions into concrete integers during this first compile:

- the FX graph is specialized to the dummy shape;
- later batches with different token or request counts trigger recompilation;
- graph cache hit rate drops;
- ACL graph capture may capture shape-specialized graphs that cannot be reused for normal serving batches;
- dynamic batching becomes less effective exactly on the hot path where vLLM needs shape flexibility.

For serving, this is not a cosmetic issue. It can cause extra compile latency, graph cache growth, and unpredictable performance when batch shape changes.

### Proposed Change.

**Meta kernels should preserve symbolic dimensions whenever output shape depends on tensor input shape.**

Use these rules in `csrc/torch_binding_meta.cpp`:

- use `tensor.sym_size(i)` instead of `tensor.size(i)` for output dimensions;
- use `tensor.sym_sizes()` instead of `tensor.sizes()` when returning same-shaped outputs;
- build shape vectors with `c10::SymDimVector` or `at::SmallVector<c10::SymInt, N>`;
- allocate outputs with `at::empty_symint(...)`;
- use `at::empty_like(...)` when the output exactly matches an input;
- keep schema parameters such as `k`, `num_heads`, and fixed constants as concrete values, wrapping them as `c10::SymInt(value)` when mixing with symbolic dimensions.

Example:

```cpp
auto batch = token_ids.sym_size(0);
c10::SymDimVector draft_shape = {batch, c10::SymInt(k)};

auto next_token_ids = at::empty_symint({batch}, token_ids.options());
auto draft_token_ids = at::empty_symint(draft_shape, token_ids.options());
```

## Checks and Guards

Meta kernels should not perform runtime validation that requires symbolic dimensions to become concrete. Avoid checks like:

```cpp
TORCH_CHECK(input.numel() > 0);
TORCH_CHECK(input.size(i) > 0);
TORCH_CHECK(input_ids.numel() == input.size(0));
```

Keep checks that only depend on non-symbolic information:

- dtype checks;
- rank checks from `dim()`;
- enum or mode argument checks;
- fixed schema parameter checks.

If a kernel truly requires a static dimension, document that requirement in the meta kernel and keep the check local. Do not silently convert a dynamic dimension to `int64_t`.

## Enforcement

This rule needs CI enforcement. Review guidelines alone are not enough because most PRs will not be reviewed by someone focused on symbolic shape behavior.

Add a small static checker, for example `tools/check_symbolic_meta.py`, scoped to `csrc/torch_binding_meta.cpp`. The checker should run in changed-lines mode in normal PR CI and full-file mode for local audits.

In changed-lines mode, reject new or modified lines in meta kernels that introduce high-risk concrete shape patterns:

```text
.size(
.sizes(
.numel(
at::empty({
torch::empty({
at::zeros({
torch::zeros({
std::vector<int64_t>
at::SmallVector<int64_t
c10::SmallVector<int64_t
```

The error message should point to the symbolic alternative:

```text
csrc/torch_binding_meta.cpp:1453 uses token_ids.size(0)
Meta kernels should preserve symbolic shape. Use token_ids.sym_size(0)
and at::empty_symint(...). If this must stay concrete, add a
symbolic-meta-ok comment with the reason.
```

Allow explicit exemptions, but require a reason:

```cpp
// symbolic-meta-ok: hidden size is a schema constant required by the kernel
auto hidden = weight.size(1);
```

The exemption is intentionally lightweight. Its purpose is to make concrete shape use visible in code review and searchable later. It should not be used for output dimensions derived from runtime batch, sequence, or token counts.

CI integration:

- add the checker to pre-commit manual hooks so `bash format.sh ci` runs it;
- run changed-lines mode in PR CI to avoid blocking on existing debt;
- keep full-file mode available for migration work and release audits.

This static guard catches the most common regressions without needing valid sample inputs for every custom op. It is not a replacement for behavior tests, but it prevents new obvious symbolic-shape breaks from silently entering the file.

### Feedback Period.

4 weeks.

### CC List.

@wangxiyuan

**Comment by yiz-liu** [comment]
For example, **with your `vllm_ascend` installed**, run the following scripts:
```python
from __future__ import annotations

from importlib import import_module

import torch
from torch import Tensor
from torch.fx import GraphModule

BATCH_SIZES = (2, 5, 9)
SEQ_LEN = 32
HIDDEN_SIZE = 128
K = 4


def bad_ngram(token_ids: Tensor, num_tokens_no_spec: Tensor) -> Tensor:
    sampled_token_ids = torch.zeros((token_ids.shape[0], 1), dtype=torch.int32)
    discard_request_mask = torch.zeros((token_ids.shape[0],), dtype=torch.bool)
    _, _, draft_token_ids, num_valid_draft_tokens = (
        torch.ops._C_ascend.npu_ngram_spec_decode(
            token_ids,
            num_tokens_no_spec,
            sampled_token_ids,
            discard_request_mask,
            32000,
            2,
            4,
            K,
        ))
    return draft_token_ids + num_valid_draft_tokens[:, None]


def good_rms_norm(x: Tensor, gamma: Tensor) -> Tensor:
    y, rstd = torch.ops._C_ascend.npu_gemma_rms_norm(x, gamma, 1e-6)
    return y + rstd


def show_meta(gm: GraphModule, op_name: str) -> None:
    for node in gm.graph.nodes:
        if node.op != "call_function" or op_name not in str(node.target):
            continue
        outputs = node.meta.get("val", node.meta.get("example_value"))
        if outputs is None:
            print(f"  {op_name} meta unavailable, keys={list(node.meta.keys())}")
            return
        if not isinstance(outputs, tuple):
            outputs = (outputs,)
        for i, out in enumerate(outputs):
            if isinstance(out, Tensor):
                shape = "[" + ", ".join(str(dim) for dim in out.shape) + "]"
                symbolic = not isinstance(out.shape[0], int)
                print(f"  output[{i}]: shape={shape}, batch_dim_symbolic={symbolic}")
        return
    print(f"  {op_name} not found in FX graph")


def run_ngram() -> None:
    compile_count = 0

    def backend(gm: GraphModule, _example_inputs: list[object]):
        nonlocal compile_count
        compile_count += 1
        print(f"\nngram compile #{compile_count}")
        show_meta(gm, "npu_ngram_spec_decode")

        def compiled(*args: object) -> Tensor:
            token_ids = next(arg for arg in args if isinstance(arg, Tensor))
            return torch.empty((token_ids.shape[0], K), dtype=token_ids.dtype)

        return compiled

    fn = torch.compile(bad_ngram, backend=backend, dynamic=True, fullgraph=True)
    for batch in BATCH_SIZES:
        fn(torch.zeros((batch, SEQ_LEN), dtype=torch.int32),
           torch.full((batch,), SEQ_LEN, dtype=torch.int32))
    print(f"ngram compiled graphs: {compile_count}")


def run_rms_norm() -> None:
    compile_count = 0

    def backend(gm: GraphModule, _example_inputs: list[object]):
        nonlocal compile_count
        compile_count += 1
        print(f"\nrms_norm compile #{compile_count}")
        show_meta(gm, "npu_gemma_rms_norm")

        def compiled(*args: object) -> Tensor:
            x = next(arg for arg in args if isinstance(arg, Tensor))
            return torch.empty_like(x)

        return compiled

    fn = torch.compile(good_rms_norm, backend=backend, dynamic=True, fullgraph=True)
    for batch in BATCH_SIZES:
        fn(torch.zeros((batch, HIDDEN_SIZE)), torch.ones((HIDDEN_SIZE,)))
    print(f"rms_norm compiled graphs: {compile_count}")


def main() -> int:
    import_module("vllm_ascend.vllm_ascend_C")
    run_ngram()
    print("\n---")
    run_rms_norm()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
```

And we should be able to observe the following outputs:

```
# python examples/sym_meta_demo.py 

ngram compile #1
  output[0]: shape=[2, 32], batch_dim_symbolic=True
  output[1]: shape=[2], batch_dim_symbolic=False
  output[2]: shape=[2, 4], batch_dim_symbolic=False
  output[3]: shape=[2], batch_dim_symbolic=False

ngram compile #2
  output[0]: shape=[5, 32], batch_dim_symbolic=True
  output[1]: shape=[5], batch_dim_symbolic=False
  output[2]: shape=[5, 4], batch_dim_symbolic=False
  output[3]: shape=[5], batch_dim_symbolic=False

ngram compile #3
  output[0]: shape=[9, 32], batch_dim_symbolic=True
  output[1]: shape=[9], batch_dim_symbolic=False
  output[2]: shape=[9, 4], batch_dim_symbolic=False
  output[3]: shape=[9], batch_dim_symbolic=False
ngram compiled graphs: 3

---

rms_norm compile #1
  output[0]: shape=[s77, s27], batch_dim_symbolic=True
  output[1]: shape=[s77, 1], batch_dim_symbolic=True
rms_norm compiled graphs: 1
```

Note that for the `rms_norm` op the shape is symbolic while for `ngram` it's not.
